server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,273 @@
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+    private static final int DEFAULT_MAP_SIZE = 32;
+    private static final int DEFAULT_LIMIT = 8;
+    private static final int MAX_LIMIT = 10_000;
+
+    private TopKAggregation(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+
+    @Nullable
+    @Override
+    public State newState(RamAccounting ramAccounting,
+                          Version indexVersionCreated,
+                          Version minNodeInCluster,
+                          MemoryManager memoryManager) {
+        return State.EMPTY;
+    }
+
+    @Override
+    public State iterate(RamAccounting ramAccounting,
+                         MemoryManager memoryManager,
+                         State state,
+                         Input<?>... args) throws CircuitBreakingException {
+        Object value = args[0].value();
+        if (state instanceof Empty) {
+            if (args.length == 2) {
+                // We have a limit provided by the user
+                Integer limit = (Integer) args[1].value();
+                if (limit <= 0 || limit > MAX_LIMIT) {
+                    throw new IllegalArgumentException(
+                        "Limit parameter for topk must be between 0 and 10_000. Got: " + limit);
+                }
+                int maxMapSize = maxMapSize(limit);
+                ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+                state = new TopKState(new ItemsSketch<>(maxMapSize), limit);
+            } else if (args.length == 1) {
+                // We don't have a limit provided, let's go with the default values
+                ramAccounting.addBytes(calculateRamUsage(DEFAULT_MAP_SIZE));
+                state = new TopKState(new ItemsSketch<>(DEFAULT_MAP_SIZE), DEFAULT_LIMIT);
+            }
+        }
+        if (state instanceof TopKState topKState) {
+            topKState.itemsSketch.update(value);
+        }
+        return state;
+    }
+
+    private static int maxMapSize(int x) {
+        // max map size should be 4 * the limit based on the power of 2 to avoid errors
+        return (int) Math.pow(2, Math.ceil(Math.log(x) / Math.log(2))) * 4;
+    }
+
+    private static long calculateRamUsage(long maxMapSize) {
+        // The internal memory space usage of item sketch will never exceed 18 * maxMapSize bytes, plus a small
+        // constant number of additional bytes.
+        // https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+        // 16 Bytes Java Object Headers for State on 64 bit
+        // 2 Bytes for Limit on 64 Bit
+        return maxMapSize * 18L + 16L + 2L;
```

Comment:
Why not use `RamUsageEstimator` for these to ensure they fit the actual JVM settings?